Edits to star-tree index page as part of site review (#199)
pinot-contrib/pinot-docs: basics/indexing/star-tree-index.md

---
description: Learn what a star-tree index is and the concept of how one works.
---

> # Star-Tree Index

Reviewer: ```suggestion
# Star-tree index
```

> In this page you will learn what a star-tree index is and gain a conceptual understanding of how one works.

Reviewer: ```suggestion
Learn what a star-tree index is and how it works.
```

Unlike other index techniques which work on a single column, the star-tree index is built on multiple columns and utilizes pre-aggregated results to significantly reduce the number of values to be processed, resulting in improved query performance.

One of the biggest challenges in realtime OLAP systems is achieving and maintaining tight SLAs on latency and throughput on large data sets. Existing techniques such as [sorted index](forward-index.md) or [inverted index](inverted-index.md) help improve query latencies, but speed-ups are still limited by the number of documents that need to be processed to compute results. On the other hand, pre-aggregating the results ensures a constant upper bound on query latencies, but can lead to storage space explosion.

> Here we introduce the **star-tree** index to utilize the pre-aggregated documents in a way that achieves low query latencies but also uses the storage space efficiently for aggregation and group-by queries.

Reviewer: ```suggestion
Use the **star-tree** index to utilize pre-aggregated documents to achieve both low query latencies and efficient use of storage space for aggregation and group-by queries.
```

{% embed url="https://www.youtube.com/watch?v=bwO0HSXguFA" %}

> ## Existing solutions

Reviewer: sounds like we're using the terms "solutions," "approaches," and "technique" interchangeably, also, the type of index is referred to as an approach....what terminology is best to use in this context?

Author: good question. as I was reading it the first time, what I thought I read was that solutions were more general, like a fix to a problem that might be taken care of using any of multiple approach options. with that, a technique would be the way an approach was applied...but I may have been reading into it.

> Consider the following data set, which is used here as an example to discuss the existing approaches:

Reviewer: Would it make sense to say "....as an example to discuss each index type as a solution?

Also, what are these indexes a solution for?

Author: I wouldn't use "solution" here. I do think it's reasonable to change this to "...to discuss these index types."

unless you go really high level (they make queries faster), the various indices aren't solving the same problem. while that makes me not like "approaches" it also makes me not like "solutions" here.

Author: ```suggestion
Consider the following data set, which is used here as an example to discuss these index types:
```

| MX      | 400         |
| USA     | 1200        |

With this approach, answering queries about total impressions for a country is a value lookup, because we have eliminated the need to process a large number of documents. However, to be able to answer queries that have multiple predicates means we would need to pre-aggregate for various combinations of different dimensions, which leads to an exponential increase in storage space.

## Star-tree solution

On one end of the spectrum we have indexing techniques that improve search times with a limited increase in space, but don't guarantee a hard upper bound on query latencies. On the other end of the spectrum, we have pre-aggregation techniques that offer a hard upper bound on query latencies, but suffer from exponential explosion of storage space

![](../../.gitbook/assets/space-time.png)

> The star-tree data structure offers a configurable trade-off between space and time and lets us achieve a hard upper bound for query latencies for a given use case. In the following sections we will define the star-tree data structure, and explain how Pinot uses it to achieve low latencies with high throughput.

Reviewer: ```suggestion
The star-tree data structure offers a configurable trade-off between space and time and lets us achieve a hard upper bound for query latencies for a given use case. The following sections cover the star-tree data structure, and explain how Pinot uses this structure to achieve low latencies with high throughput.
```

* **Root node** (Orange): Single root node, from which the rest of the tree can be traversed.
* **Leaf node** (Blue): A leaf node can containing at most _T_ records, where _T_ is configurable.
* **Non-leaf node** (Green): Nodes with more than _T_ records are further split into children nodes.

> * **Star node** (Yellow): Non-leaf nodes can also have a special child node called the star node. This node contains the pre-aggregated records after removing the dimension on which the data was split for this level.

Reviewer: ```suggestion
* **Star node** (Yellow): Non-leaf nodes can have a special child node called the star node. This node contains the pre-aggregated records after removing the dimension on which the data was split for this level.
```

  * If a node has more than _T_ records, it is split into multiple children nodes, one for each value of the dimension in the split order corresponding to current level in the tree.
  *   A star node can be created (per configuration) for the current node, by dropping the dimension being split on, and aggregating the metrics for rows containing dimensions with identical values. These aggregated documents are appended to the end of the star-tree documents.

>       If there is only one value for the current dimension, a star node won’t be created because the documents under the star node are identical to the single node.

Reviewer: ```suggestion
If the current dimension only has one value, a star node isn't created because its documents would be identical to the single node.
```

* The above step is repeated recursively until there are no more nodes to split.

> * Multiple star-trees can be generated based on different configurations (_dimensionsSplitOrder_, _aggregations_, _T_)

Reviewer: ```suggestion
* Multiple star-tree nodes can be generated based on different configurations (_dimensionsSplitOrder_, _aggregations_, _T_)
```

The star-tree and documents should be something like below:

> ### **Tree structure**

Reviewer: ```suggestion
### Tree structure
```
